rune/libenclave/internal/runtime/pal/skeleton/liberpal-skeleton.c

```diff
@@ -173,9 +174,43 @@ static bool encl_create(int dev_fd, unsigned long bin_size,
 	secs->attributes = SGX_ATTR_MODE64BIT;
 	if (enclave_debug)
 		secs->attributes |= SGX_ATTR_DEBUG;
+	if (meta_data->attributes) {
+		/* Check attributes to prevent possible tampering of metadata area. */
+		if ((meta_data->attributes & sigstruct->body.attributes) !=
+		    sigstruct->body.attributes) {
+			fprintf(stderr, "Invalid attributes value.\n");
+			return false;
+		}
+
+		secs->attributes = meta_data->attributes & secs->attributes;
+	}
+	/* Check the attributes in signature structure restrictions. */
+	if ((sigstruct->body.attributes & sigstruct->body.attributes_mask) !=
+	    (secs->attributes & sigstruct->body.attributes_mask)) {
+		fprintf(stderr,
+			"secs attributes does NOT match signature attributes.\n");
+		return false;
+	}
 
 	get_sgx_xfrm_by_cpuid(&xfrm);
 	secs->xfrm = xfrm;
+	if (meta_data->xfrm) {
+		/* Check whether the xfrm features in metadata area are available. */
+		if ((secs->xfrm & meta_data->xfrm) != meta_data->xfrm) {
+			fprintf(stderr, "Invalid xfrm value. "
+				"Supported xfrm features are %#lx.\n",
+				secs->xfrm);
```

Comment:
You need to report the unavailable bits claimed in meta_data->xfrm.

Reply:
Use `meta_data->xfrm & ~(secs->xfrm & meta_data->xfrm)` to get the mask of unavailable bits.